processing/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -52,333 +56,313 @@
-  // This is a lock against the state of the REFERENCE MANAGER (aka start/stop state), NOT of the lookup itself.
-  private final ReadWriteLock startStopLock = new ReentrantReadWriteLock(true);
-  private final AtomicBoolean started = new AtomicBoolean(false);
+  private static final EmittingLogger LOG = new EmittingLogger(LookupReferencesManager.class);
+
+  private final Map<String, LookupExtractorFactoryContainer> lookupMap = new ConcurrentHashMap<>();
+
+  @VisibleForTesting
+  final BlockingQueue<Notice> queue = new LinkedBlockingDeque<>(10000);
 
   private final LookupSnapshotTaker lookupSnapshotTaker;
 
+  @VisibleForTesting
+  final LifecycleLock lifecycleLock = new LifecycleLock();
+
+  @VisibleForTesting
+  Thread mainThread;
+
+  //for unit testing only
+  private final boolean testMode;
+
   @Inject
-  public LookupReferencesManager(LookupConfig lookupConfig, final @Json ObjectMapper objectMapper)
+  public LookupReferencesManager(LookupConfig lookupConfig, @Json ObjectMapper objectMapper)
+  {
+    this(lookupConfig, objectMapper, false);
+  }
+
+  @VisibleForTesting
+  LookupReferencesManager(LookupConfig lookupConfig, ObjectMapper objectMapper, boolean testMode)
   {
     if (Strings.isNullOrEmpty(lookupConfig.getSnapshotWorkingDir())) {
       this.lookupSnapshotTaker = null;
     } else {
       this.lookupSnapshotTaker = new LookupSnapshotTaker(objectMapper, lookupConfig.getSnapshotWorkingDir());
     }
+
+    this.testMode = testMode;
   }
 
   @LifecycleStart
   public void start()
   {
-    startStopLock.writeLock().lock();
     try {
-      if (!started.getAndSet(true)) {
-        if (lookupSnapshotTaker != null) {
-          final List<LookupBean> lookupBeanList = lookupSnapshotTaker.pullExistingSnapshot();
-          for (LookupBean lookupBean : lookupBeanList) {
-            this.put(lookupBean.name, lookupBean.factory);
+      if (!lifecycleLock.canStart()) {
+        throw new ISE("can't start.");
+      }
+
+      LOG.info("LookupReferencesManager is starting.");
+
+      if (lookupSnapshotTaker != null) {
+        final List<LookupBean> lookupBeanList = lookupSnapshotTaker.pullExistingSnapshot();
+        for (LookupBean lookupBean : lookupBeanList) {
+          LookupExtractorFactoryContainer container = lookupBean.getContainer();
+
+          if (container.getLookupExtractorFactory().start()) {
+            lookupMap.put(lookupBean.getName(), container);
+          } else {
+            throw new ISE("Failed to start lookup [%s]:[%s]", lookupBean.getName(), container);
           }
         }
-        LOGGER.info("Started lookup factory references manager");
       }
+
+      if (!testMode) {
+        mainThread = new Thread(
+            new Runnable()
+            {
+              @Override
+              public void run()
+              {
+                try {
+                  while (lifecycleLock.isStarted() && !Thread.interrupted()) {
```

Comment:
isStarted() shouldn't be checked on each iteration, could be checked just once in the beginning. The "exit synchronization" (_no longer_ isStarted()) is ensured by mainThread.join() in stop().